Automattic/wp-calypso
# Payments: Wrap the account-level change payment method in an error boundary

#### Changes proposed in this Pull Request

While the site-level "Change payment method" page [is wrapped in a React error boundary](https://github.com/Automattic/wp-calypso/blob/10dfe28aed202372f43f7bd5c1cfcf628d54d289/client/my-sites/purchases/main.tsx#L190) to catch fatal errors and report them to the payments team, the account-level page is not.

This PR adds an error boundary to the account-level page. Note that the "Change payment method" form itself _is_ wrapped in an error boundary by virtue of being inside a [CheckoutProvider](https://github.com/Automattic/wp-calypso/blob/10dfe28aed202372f43f7bd5c1cfcf628d54d289/client/me/purchases/manage-purchase/payment-method-selector/index.tsx#L95) (although those errors are not being logged; that's [a separate issue](https://github.com/Automattic/wp-calypso/issues/57088)), but the components above that point are not.

This also re-uses the error boundary class from the `composite-checkout` package for all of payments, since it is identical to the one that was already being used.

Inspired by the fatal errors reported in https://github.com/Automattic/wp-calypso/issues/57087. This will not actually affect that case because the thrown error is happening in the notices module, well above the error boundary.

<img width="1127" alt="Screen Shot 2021-10-18 at 11 55 59 AM" src="https://user-images.githubusercontent.com/2036909/137767383-aaa749e2-b358-43d1-b8a6-1f9517761039.png">

#### Testing instructions

- Visit the site-level and the account-level "change payment method" page and verify that everything renders as expected. The site-level one can be found by visiting `/purchases/YOURSITEHERE`, clicking on an active subscription, then clicking the "Change payment method" button at the bottom. The account-level one can be found by visiting `/me/purchases` and repeating the same steps.

Next, cause an error by applying the following diff on top of this PR:

```diff
diff --git a/client/me/purchases/manage-purchase/change-payment-method/index.tsx b/client/me/purchases/manage-purchase/change-payment-method/index.tsx
index 47b47aba36..d54cbb8aae 100644
--- a/client/me/purchases/manage-purchase/change-payment-method/index.tsx
+++ b/client/me/purchases/manage-purchase/change-payment-method/index.tsx
@@ -57,6 +57,7 @@ function ChangePaymentMethod( {
        const isDataLoading =
                ! hasLoadedSites || ! hasLoadedUserPurchases || ! hasLoadedStoredCards || isStripeLoading;
        const isDataValid = purchase && selectedSite;
+       throw new Error( 'testing' );

        useEffect( () => {
                if ( ! isDataLoading && ! isDataValid ) {
```

Then view the account-level "Change payment method" page and verify that you see the error in the screenshot above.
